test/e2e/scenario/recorder/viewports.scenario.ts

```diff
@@ -0,0 +1,371 @@
+import { IncrementalSource, ViewportResizeData, ScrollData } from '@datadog/browser-rum/cjs/types'
+import { RumInitConfiguration } from '@datadog/browser-rum-core'
+
+import { findAllIncrementalSnapshots, findAllVisualViewports } from '@datadog/browser-rum/test/utils'
+import { createTest, bundleSetup, html, EventRegistry } from '../../lib/framework'
+import { browserExecute } from '../../lib/helpers/browser'
+import { flushEvents } from '../../lib/helpers/sdk'
+
+const VIEWPORT_META_TAGS = `
+<meta name="apple-mobile-web-app-capable" content="yes">
+<meta name="mobile-web-app-capable" content="yes">
+<meta name="viewport"
+  content="width=device-width, initial-scale=1.0, maximum-scale=2.75, minimum-scale=1.0, user-scalable=yes"
+>
+`
+
+describe('recorder', () => {
+  beforeEach(() => {
+    if (isGestureUnsupported()) {
+      pending('no touch gesture support')
+    }
+  })
+
+  describe('layout viewport properties', () => {
+    createTest('getWindowWidth/Height should not be affected by pinch zoom')
+      .withRum({ enableExperimentalFeatures: ['visualviewport'] })
+      .withRumInit(initRumAndStartRecording)
+      .withSetup(bundleSetup)
+      .withBody(html`${VIEWPORT_META_TAGS}`)
+      .run(async ({ events }) => {
+        await buildScrollablePage()
+
+        const { innerWidth, innerHeight } = await getWindowInnerDimensions()
+        await performSignificantZoom()
+
+        await browserExecute(() => {
+          window.dispatchEvent(new Event('resize'))
+        })
+
+        const lastViewportResizeData = (
+          await getLastRecord(events, (segment) =>
+            findAllIncrementalSnapshots(segment, IncrementalSource.ViewportResize)
+          )
+        ).data as ViewportResizeData
+
+        const scrollbarWidth = await getScrollbarCorrection()
+
+        expectToBeNearby(lastViewportResizeData.width, innerWidth - scrollbarWidth)
+        expectToBeNearby(lastViewportResizeData.height, innerHeight - scrollbarWidth)
+      })
+
+    /**
+     * window.ScrollX/Y on some devices/browsers are changed by pinch zoom
+     * We need to ensure that our measurements are not affected by pinch zoom
+     */
+    createTest('getScrollX/Y should not be affected by pinch scroll')
+      .withRum({ enableExperimentalFeatures: ['visualviewport'] })
+      .withRumInit(initRumAndStartRecording)
+      .withSetup(bundleSetup)
+      .withBody(html`${VIEWPORT_META_TAGS}`)
+      .run(async ({ events }) => {
+        const VISUAL_SCROLL_DOWN_PX = 60
+        const LAYOUT_SCROLL_AMOUNT = 20
+
+        await buildScrollablePage()
+        await performSignificantZoom()
+        await resetLayoutScroll()
+
+        const initialVisualViewport = await getVisualViewport()
+        const { scrollX: initialScrollX, scrollY: initialScrollY } = await getWindowScroll()
+
+        // Add Visual Viewport Scroll
+        await pinchScrollVerticallyDown(VISUAL_SCROLL_DOWN_PX)
```

Comment:
why not having a `visualScrollTo` helper to be symetric with `layoutScrollTo`?

Reply:
Unfortunately `layoutScrollTo` uses the native JS method to scroll to an absolute position whereas `pinchScrollVerticallyDown` uses the browser automation API to relatively pinch.
Unfortunately the nature of pinchScroll means that the direction is poorly defined (iOS vs Android) and that pinch scrolling too much will "overflow" into the layout scroll.

Renaming `pinchScrollVerticallyDown` to `visualScrollVerticallyDown` might be more consistent, but it wouldn't be perfectly accurate either.